javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,373 @@
+
+  override def clear(): Unit = {
+    head = null
+    last = null
+    _size = 0
+  }
+
+  private def getNodeAt(index: Int): Node[E] = {
+    if (index == 0) head
+    else if (index == size - 1) last
+    else {
+      var current: Node[E] = null
+      if (index <= size/2) {
+        current = head
+        for (_ <- 0 until index) current = current.next
+      } else {
+        current = last
+        for (_ <- index until (size - 1)) current = current.prev
+      }
+      current
+    }
+  }
+
+  override def get(index: Int): E = {
+    checkIndexInBounds(index)
+    getNodeAt(index).value
+  }
+
+  override def set(index: Int, element: E): E = {
+    checkIndexInBounds(index)
+    val node = getNodeAt(index)
+    val oldValue = node.value
+    node.value = element
+    oldValue
+  }
+
+  private def addNode(nextNode: Node[E], e: E): Unit = {
+    if (nextNode eq head) addFirst(e)
+    else if (nextNode eq null) addLast(e)
+    else {
+
+      val node = new Node(e, prev = nextNode.prev, next = nextNode)
+      nextNode.prev.next = node
+      nextNode.prev = node
+
+      _size += 1
+    }
+  }
+
+  override def add(index: Int, element: E): Unit = {
+    checkIndexOnBounds(index)
+    addNode(getNodeAt(index), element)
+  }
+
+  private def removeNode(node: Node[E]): E = {
+    if (node eq head) removeFirst()
+    else if (node eq last) removeLast()
+    else {
+
+      node.prev.next = node.next
+      node.next.prev = node.prev
+
+      _size -= 1
+
+      node.value
+    }
+  }
+
+  override def remove(index: Int): E = {
+    checkIndexInBounds(index)
+    removeNode(getNodeAt(index))
+  }
+
+  def peek(): E =
+    peekFirst()
+
+  def element(): E =
+    getFirst()
+
+  def poll(): E =
+    pollFirst()
+
+  def remove(): E =
+    removeFirst()
+
+  def offer(e: E): Boolean =
+    offerLast(e)
+
+  def offerFirst(e: E): Boolean = {
+    addFirst(e)
+    true
+  }
+
+  def offerLast(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  def peekFirst(): E =
+    head.value
+
+  def peekLast(): E =
+    peekFirst()
+
+  def pollFirst(): E =
+    removeFirst()
+
+  def pollLast(): E =
+    removeLast()
+
+  def push(e: E): Unit =
+    addFirst(e)
+
+  def pop(): E =
+    removeFirst()
+
+  private def _removeOccurrence(iter: Iterator[E], o: Any): Boolean = {
+    var changed = false
+    val e = o.asInstanceOf[E]
+    while (iter.hasNext() && !changed) {
+      if (iter.next() === e) {
+        iter.remove()
+        changed = true
+      }
+    }
+
+    changed
+  }
+
+  def removeFirstOccurrence(o: Any): Boolean =
+    _removeOccurrence(iterator(), o)
+
+  def removeLastOccurrence(o: Any): Boolean =
+    _removeOccurrence(descendingIterator(), o)
+
+  override def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    new ListIterator[E] {
+
+      private var last = -1
+      private var i = index
+
+      private var currentNode: Node[E] =
+        if (index == size) null
+        else getNodeAt(index)
```

Comment:
going previous when the current node is the first element should throw in the first place (since there is no previous element to return) so that cannot happen. For example:
```scala
val iter = list.listIterator()
// currentNode points to the first node of the list
expect(iter.nextIndex).toEqual(0)
iter.previous() // throws, doesn't change the state
// currentNode still points to the first node of the list
```